# **Part 1: Loading & Basic Information.**

**1. Load the Iris dataset using pd.read_csv()**

In [4]:
import pandas as pd
csv = pd.read_csv('Iris.csv')
csv

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


**2. Display**

In [7]:
csv.head(6) #First 5 rows

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
5,6,5.4,3.9,1.7,0.4,Iris-setosa


In [9]:
csv.tail(5) #Last 5 rows

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica
149,150,5.9,3.0,5.1,1.8,Iris-virginica


In [10]:
csv.sample(n=10) #10 random rows

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
98,99,5.1,2.5,3.0,1.1,Iris-versicolor
13,14,4.3,3.0,1.1,0.1,Iris-setosa
81,82,5.5,2.4,3.7,1.0,Iris-versicolor
85,86,6.0,3.4,4.5,1.6,Iris-versicolor
148,149,6.2,3.4,5.4,2.3,Iris-virginica
138,139,6.0,3.0,4.8,1.8,Iris-virginica
60,61,5.0,2.0,3.5,1.0,Iris-versicolor
102,103,7.1,3.0,5.9,2.1,Iris-virginica
48,49,5.3,3.7,1.5,0.2,Iris-setosa
62,63,6.0,2.2,4.0,1.0,Iris-versicolor


**3. Find**

In [17]:
print(len(csv.columns))  ##Number of columns

6


In [18]:
print(csv.shape) #shape

(150, 6)


In [22]:
print(csv.dtypes) #datatypes

Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object


In [31]:
print(csv.memory_usage(deep=True))  #memory usage

Index             132
Id               1200
SepalLengthCm    1200
SepalWidthCm     1200
PetalLengthCm    1200
PetalWidthCm     1200
Species          9350
dtype: int64


In [32]:
print(csv.columns) #columns name

Index(['Id', 'SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm',
       'Species'],
      dtype='object')


 **4. Use**

In [33]:
csv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             150 non-null    int64  
 1   SepalLengthCm  150 non-null    float64
 2   SepalWidthCm   150 non-null    float64
 3   PetalLengthCm  150 non-null    float64
 4   PetalWidthCm   150 non-null    float64
 5   Species        150 non-null    object 
dtypes: float64(4), int64(1), object(1)
memory usage: 7.2+ KB


In [34]:
csv.describe()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
count,150.000000,150.000000,150.000000,150.000000,150.000000
mean,75.500000,5.843333,3.054000,3.758667,1.198667
std,43.445368,0.828066,0.433594,1.764420,0.763161
min,1.000000,4.300000,2.000000,1.000000,0.100000
25%,38.250000,5.100000,2.800000,1.600000,0.300000
50%,75.500000,5.800000,3.000000,4.350000,1.300000
75%,112.750000,6.400000,3.300000,5.100000,1.800000
max,150.000000,7.900000,4.400000,6.900000,2.500000


# **Part 2: Selecting Data**

**5. Print only**

In [36]:
print(csv["SepalLengthCm"])

0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: SepalLengthCm, Length: 150, dtype: float64


In [37]:
print(csv["PetalLengthCm"])

0      1.4
1      1.4
2      1.3
3      1.5
4      1.4
      ... 
145    5.2
146    5.0
147    5.2
148    5.4
149    5.1
Name: PetalLengthCm, Length: 150, dtype: float64


**6. Print**

In [38]:
print(csv[:10]) #First 10 rows

   Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa
5   6            5.4           3.9            1.7           0.4  Iris-setosa
6   7            4.6           3.4            1.4           0.3  Iris-setosa
7   8            5.0           3.4            1.5           0.2  Iris-setosa
8   9            4.4           2.9            1.4           0.2  Iris-setosa
9  10            4.9           3.1            1.5           0.1  Iris-setosa


In [39]:
print(csv[-11:-1])  #Last 10 rows

      Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm  \
139  140            6.9           3.1            5.4           2.1   
140  141            6.7           3.1            5.6           2.4   
141  142            6.9           3.1            5.1           2.3   
142  143            5.8           2.7            5.1           1.9   
143  144            6.8           3.2            5.9           2.3   
144  145            6.7           3.3            5.7           2.5   
145  146            6.7           3.0            5.2           2.3   
146  147            6.3           2.5            5.0           1.9   
147  148            6.5           3.0            5.2           2.0   
148  149            6.2           3.4            5.4           2.3   

            Species  
139  Iris-virginica  
140  Iris-virginica  
141  Iris-virginica  
142  Iris-virginica  
143  Iris-virginica  
144  Iris-virginica  
145  Iris-virginica  
146  Iris-virginica  
147  Iris-virginica  
148  Iris

In [40]:
print(csv[20:41]) #Rows 20–40

    Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
20  21            5.4           3.4            1.7           0.2  Iris-setosa
21  22            5.1           3.7            1.5           0.4  Iris-setosa
22  23            4.6           3.6            1.0           0.2  Iris-setosa
23  24            5.1           3.3            1.7           0.5  Iris-setosa
24  25            4.8           3.4            1.9           0.2  Iris-setosa
25  26            5.0           3.0            1.6           0.2  Iris-setosa
26  27            5.0           3.4            1.6           0.4  Iris-setosa
27  28            5.2           3.5            1.5           0.2  Iris-setosa
28  29            5.2           3.4            1.4           0.2  Iris-setosa
29  30            4.7           3.2            1.6           0.2  Iris-setosa
30  31            4.8           3.1            1.6           0.2  Iris-setosa
31  32            5.4           3.4            1.5           0.4

**7. Using iloc**

In [50]:
csv.iloc[:5,:3] #First 5 rows and first 3 columns

,Id,SepalLengthCm,SepalWidthCm
0,1,5.1,3.5
1,2,4.9,3.0
2,3,4.7,3.2
3,4,4.6,3.1
4,5,5.0,3.6


In [52]:
csv.iloc[-11:-1] #Last 10 rows

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
139,140,6.9,3.1,5.4,2.1,Iris-virginica
140,141,6.7,3.1,5.6,2.4,Iris-virginica
141,142,6.9,3.1,5.1,2.3,Iris-virginica
142,143,5.8,2.7,5.1,1.9,Iris-virginica
143,144,6.8,3.2,5.9,2.3,Iris-virginica
144,145,6.7,3.3,5.7,2.5,Iris-virginica
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


**8. Using loc**

In [62]:
csv.loc[csv["Species"] == "Iris-setosa"]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
5,6,5.4,3.9,1.7,0.4,Iris-setosa
6,7,4.6,3.4,1.4,0.3,Iris-setosa
7,8,5.0,3.4,1.5,0.2,Iris-setosa
8,9,4.4,2.9,1.4,0.2,Iris-setosa
9,10,4.9,3.1,1.5,0.1,Iris-setosa


# **Part 3: Filtering**

**9. Flowers with:**

In [63]:
csv.loc[csv["SepalLengthCm"] > 5]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
5,6,5.4,3.9,1.7,0.4,Iris-setosa
10,11,5.4,3.7,1.5,0.2,Iris-setosa
14,15,5.8,4.0,1.2,0.2,Iris-setosa
15,16,5.7,4.4,1.5,0.4,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


**10. Flowers with:**

In [64]:
csv.loc[csv["PetalLengthCm"] > 4]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
50,51,7.0,3.2,4.7,1.4,Iris-versicolor
51,52,6.4,3.2,4.5,1.5,Iris-versicolor
52,53,6.9,3.1,4.9,1.5,Iris-versicolor
54,55,6.5,2.8,4.6,1.5,Iris-versicolor
55,56,5.7,2.8,4.5,1.3,Iris-versicolor
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


**11. Flowers with:**

In [65]:
csv.loc[csv["PetalWidthCm"] > 1.5]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
56,57,6.3,3.3,4.7,1.6,Iris-versicolor
70,71,5.9,3.2,4.8,1.8,Iris-versicolor
77,78,6.7,3.0,5.0,1.7,Iris-versicolor
83,84,6.0,2.7,5.1,1.6,Iris-versicolor
85,86,6.0,3.4,4.5,1.6,Iris-versicolor
100,101,6.3,3.3,6.0,2.5,Iris-virginica
101,102,5.8,2.7,5.1,1.9,Iris-virginica
102,103,7.1,3.0,5.9,2.1,Iris-virginica
103,104,6.3,2.9,5.6,1.8,Iris-virginica
104,105,6.5,3.0,5.8,2.2,Iris-virginica


 **12. Flowers with:**

In [68]:
csv.loc[(csv['SepalLengthCm'] > 5) & (csv["PetalWidthCm"] > 1)]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
50,51,7.0,3.2,4.7,1.4,Iris-versicolor
51,52,6.4,3.2,4.5,1.5,Iris-versicolor
52,53,6.9,3.1,4.9,1.5,Iris-versicolor
53,54,5.5,2.3,4.0,1.3,Iris-versicolor
54,55,6.5,2.8,4.6,1.5,Iris-versicolor
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


 **13. Flowers with:**

In [69]:
csv.loc[(csv['SepalWidthCm'] < 3) & (csv["Species"] == "Iris-versicolor")]

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
53,54,5.5,2.3,4.0,1.3,Iris-versicolor
54,55,6.5,2.8,4.6,1.5,Iris-versicolor
55,56,5.7,2.8,4.5,1.3,Iris-versicolor
57,58,4.9,2.4,3.3,1.0,Iris-versicolor
58,59,6.6,2.9,4.6,1.3,Iris-versicolor
59,60,5.2,2.7,3.9,1.4,Iris-versicolor
60,61,5.0,2.0,3.5,1.0,Iris-versicolor
62,63,6.0,2.2,4.0,1.0,Iris-versicolor
63,64,6.1,2.9,4.7,1.4,Iris-versicolor
64,65,5.6,2.9,3.6,1.3,Iris-versicolor


# **Part 4: Sorting**

**14. Sort by:
Sepal Length (Ascending)**

In [73]:
csv.sort_values(by="SepalLengthCm", ascending=True)

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
13,14,4.3,3.0,1.1,0.1,Iris-setosa
8,9,4.4,2.9,1.4,0.2,Iris-setosa
42,43,4.4,3.2,1.3,0.2,Iris-setosa
38,39,4.4,3.0,1.3,0.2,Iris-setosa
41,42,4.5,2.3,1.3,0.3,Iris-setosa
...,...,...,...,...,...,...
122,123,7.7,2.8,6.7,2.0,Iris-virginica
117,118,7.7,3.8,6.7,2.2,Iris-virginica
118,119,7.7,2.6,6.9,2.3,Iris-virginica
135,136,7.7,3.0,6.1,2.3,Iris-virginica


**15. Sort by:
Petal Length (Descending)**

In [74]:
csv.sort_values(by="PetalLengthCm", ascending=False)

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
118,119,7.7,2.6,6.9,2.3,Iris-virginica
117,118,7.7,3.8,6.7,2.2,Iris-virginica
122,123,7.7,2.8,6.7,2.0,Iris-virginica
105,106,7.6,3.0,6.6,2.1,Iris-virginica
131,132,7.9,3.8,6.4,2.0,Iris-virginica
...,...,...,...,...,...,...
41,42,4.5,2.3,1.3,0.3,Iris-setosa
35,36,5.0,3.2,1.2,0.2,Iris-setosa
14,15,5.8,4.0,1.2,0.2,Iris-setosa
13,14,4.3,3.0,1.1,0.1,Iris-setosa


**16. Sort by:**
**1.Species**
**2.Petal Length**

In [75]:
csv.sort_values(by=["Species", "PetalLengthCm"])

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
22,23,4.6,3.6,1.0,0.2,Iris-setosa
13,14,4.3,3.0,1.1,0.1,Iris-setosa
14,15,5.8,4.0,1.2,0.2,Iris-setosa
35,36,5.0,3.2,1.2,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
...,...,...,...,...,...,...
131,132,7.9,3.8,6.4,2.0,Iris-virginica
105,106,7.6,3.0,6.6,2.1,Iris-virginica
117,118,7.7,3.8,6.7,2.2,Iris-virginica
122,123,7.7,2.8,6.7,2.0,Iris-virginica


# **Part 5: Statistics**

**17.**
* **Mean**
* **Median**
* **Standard Deviation**
* **Variance**
* **Maximum**
* **Minimum**













In [77]:
#SepalLengthCm
print("mean:",csv["SepalLengthCm"].mean())
print("median:",csv["SepalLengthCm"].median())
print("std:",csv["SepalLengthCm"].std())
print("var:",csv["SepalLengthCm"].var())
print("max:",csv["SepalLengthCm"].max())
print("min:",csv["SepalLengthCm"].min())

mean: 5.843333333333334
median: 5.8
std: 0.8280661279778629
var: 0.6856935123042505
max: 7.9
min: 4.3


In [78]:
#SepalWidthCm
print("mean:",csv["SepalWidthCm"].mean())
print("median:",csv["SepalWidthCm"].median())
print("std:",csv["SepalWidthCm"].std())
print("var:",csv["SepalWidthCm"].var())
print("max:",csv["SepalWidthCm"].max())
print("min:",csv["SepalWidthCm"].min())

mean: 3.0540000000000003
median: 3.0
std: 0.4335943113621737
var: 0.18800402684563763
max: 4.4
min: 2.0


In [79]:
#PetalLengthCm
print("mean:",csv["PetalLengthCm"].mean())
print("median:",csv["PetalLengthCm"].median())
print("std:",csv["PetalLengthCm"].std())
print("var:",csv["PetalLengthCm"].var())
print("max:",csv["PetalLengthCm"].max())
print("min:",csv["PetalLengthCm"].min())

mean: 3.758666666666666
median: 4.35
std: 1.7644204199522617
var: 3.1131794183445156
max: 6.9
min: 1.0


In [80]:
#PetalWidthCm
print("mean:",csv["PetalWidthCm"].mean())
print("median:",csv["PetalWidthCm"].median())
print("std:",csv["PetalWidthCm"].std())
print("var:",csv["PetalWidthCm"].var())
print("max:",csv["PetalWidthCm"].max())
print("min:",csv["PetalWidthCm"].min())

mean: 1.1986666666666668
median: 1.3
std: 0.7631607417008414
var: 0.5824143176733784
max: 2.5
min: 0.1


# **Part 6: Unique Values**

**18. Find:
Unique species**

In [81]:
csv["Species"].unique()

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

**19. Count flowers of each species**

In [87]:
print(csv.loc[csv["Species"] == "Iris-setosa"].count())
print(csv.loc[csv["Species"] == "Iris-versicolor"].count())
print(csv.loc[csv["Species"] == "Iris-virginica"].count())

Id               50
SepalLengthCm    50
SepalWidthCm     50
PetalLengthCm    50
PetalWidthCm     50
Species          50
dtype: int64
Id               50
SepalLengthCm    50
SepalWidthCm     50
PetalLengthCm    50
PetalWidthCm     50
Species          50
dtype: int64
Id               50
SepalLengthCm    50
SepalWidthCm     50
PetalLengthCm    50
PetalWidthCm     50
Species          50
dtype: int64


**20. Which species appears the most?**

In [88]:
csv["Species"].value_counts()

,count
Species,
Iris-setosa,50
Iris-versicolor,50
Iris-virginica,50


# **Part 7: Missing Values**

**21. Check missing values.**

In [90]:
csv.isnull()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,False,False,False,False,False,False
1,False,False,False,False,False,False
2,False,False,False,False,False,False
3,False,False,False,False,False,False
4,False,False,False,False,False,False
...,...,...,...,...,...,...
145,False,False,False,False,False,False
146,False,False,False,False,False,False
147,False,False,False,False,False,False
148,False,False,False,False,False,False


**22. Find missing values column-wise.**

In [91]:
csv.isnull().sum()

,0
Id,0
SepalLengthCm,0
SepalWidthCm,0
PetalLengthCm,0
PetalWidthCm,0
Species,0


**23. Fill missing values with:**
* Mean
* Median

In [92]:
csv.fillna(csv.mean(numeric_only=True))

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


In [93]:
csv.fillna(csv.median(numeric_only=True))

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


**24. Remove rows with missing values.**

In [94]:
csv.dropna()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species
0,1,5.1,3.5,1.4,0.2,Iris-setosa
1,2,4.9,3.0,1.4,0.2,Iris-setosa
2,3,4.7,3.2,1.3,0.2,Iris-setosa
3,4,4.6,3.1,1.5,0.2,Iris-setosa
4,5,5.0,3.6,1.4,0.2,Iris-setosa
...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica
146,147,6.3,2.5,5.0,1.9,Iris-virginica
147,148,6.5,3.0,5.2,2.0,Iris-virginica
148,149,6.2,3.4,5.4,2.3,Iris-virginica


# **Part 8: GroupBy**

**25. Find average:**
* **Sepal Length**
* **Petal Length**
* **Petal Width**

**for each species.**

In [96]:
csv.groupby("Species")["SepalLengthCm"].mean()

,SepalLengthCm
Species,
Iris-setosa,5.006
Iris-versicolor,5.936
Iris-virginica,6.588


In [97]:
csv.groupby("Species")["PetalLengthCm"].mean()

,PetalLengthCm
Species,
Iris-setosa,1.464
Iris-versicolor,4.260
Iris-virginica,5.552


In [98]:
csv.groupby("Species")["PetalWidthCm"].mean()

,PetalWidthCm
Species,
Iris-setosa,0.244
Iris-versicolor,1.326
Iris-virginica,2.026


**26. Find maximum values for each species.**

In [99]:
csv.groupby("Species").max()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
Species,,,,,
Iris-setosa,50,5.8,4.4,1.9,0.6
Iris-versicolor,100,7.0,3.4,5.1,1.8
Iris-virginica,150,7.9,3.8,6.9,2.5


**27. Count flowers in each species using groupby().**

In [101]:
csv.groupby("Species").count()

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm
Species,,,,,
Iris-setosa,50,50,50,50,50
Iris-versicolor,50,50,50,50,50
Iris-virginica,50,50,50,50,50


# **Part 9: Apply Function**

**28. Create a new column:**
**Flower Size**

**Rule:**

* **Petal Length > 4 → Large**
* **Otherwise → Small**

**Use apply().**

In [103]:
def size(PetalLengthCm):
  if PetalLengthCm > 4:
    return "Large"
  else:
    return "Small"
csv["Flower Size"] = csv["PetalLengthCm"].apply(size)
csv

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species,Flower Size
0,1,5.1,3.5,1.4,0.2,Iris-setosa,Small
1,2,4.9,3.0,1.4,0.2,Iris-setosa,Small
2,3,4.7,3.2,1.3,0.2,Iris-setosa,Small
3,4,4.6,3.1,1.5,0.2,Iris-setosa,Small
4,5,5.0,3.6,1.4,0.2,Iris-setosa,Small
...,...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica,Large
146,147,6.3,2.5,5.0,1.9,Iris-virginica,Large
147,148,6.5,3.0,5.2,2.0,Iris-virginica,Large
148,149,6.2,3.4,5.4,2.3,Iris-virginica,Large


**29. Create another column:
Sepal Area**

**Formula:**

**SepalLengthCm × SepalWidthCm**

In [109]:
csv["Sepal Area"] = csv.apply(lambda x: x["SepalLengthCm"] * x["SepalWidthCm"],axis = 1)
csv

,Id,SepalLengthCm,SepalWidthCm,PetalLengthCm,PetalWidthCm,Species,Flower Size,Sepal Area
0,1,5.1,3.5,1.4,0.2,Iris-setosa,Small,17.85
1,2,4.9,3.0,1.4,0.2,Iris-setosa,Small,14.70
2,3,4.7,3.2,1.3,0.2,Iris-setosa,Small,15.04
3,4,4.6,3.1,1.5,0.2,Iris-setosa,Small,14.26
4,5,5.0,3.6,1.4,0.2,Iris-setosa,Small,18.00
...,...,...,...,...,...,...,...,...
145,146,6.7,3.0,5.2,2.3,Iris-virginica,Large,20.10
146,147,6.3,2.5,5.0,1.9,Iris-virginica,Large,15.75
147,148,6.5,3.0,5.2,2.0,Iris-virginica,Large,19.50
148,149,6.2,3.4,5.4,2.3,Iris-virginica,Large,21.08


**Part 10: Merge**

**30. Create:**

In [111]:
df1 = pd.DataFrame({"ID":[1,2,3],"Name":["A","B","C"]})
df2 = pd.DataFrame({"ID":[1,2,3],"Marks":["85","90","75"]})
pd.merge(df1,df2,on="ID")


,ID,Name,Marks
0,1,A,85
1,2,B,90
2,3,C,75


## **Part 11: Concat**
* Concatenate vertically.
 * Concatenate horizontally.

In [112]:
pd.concat([df1,df2],axis=0)

,ID,Name,Marks
0,1,A,NaN
1,2,B,NaN
2,3,C,NaN
0,1,NaN,85
1,2,NaN,90
2,3,NaN,75


In [113]:
pd.concat([df1,df2],axis=1)

,ID,Name,ID,Marks
0,1,A,1,85
1,2,B,2,90
2,3,C,3,75


# **Part 12: Pivot Table**

**Create a pivot table showing:**

* **Index → Species**
* **Values → Petal Length**
* **Aggregation → Mean**

In [115]:
csv.pivot_table(index="Species",values="PetalLengthCm",aggfunc="mean")

,PetalLengthCm
Species,
Iris-setosa,1.464
Iris-versicolor,4.260
Iris-virginica,5.552


# **Analysis**
* **Which species has the highest average petal length?**
* *Ans: Iris-virginica*

* **Which species has the widest sepals?**
* *Ans:Iris-setosa*

* **Which species occurs the most?**
* *Ans: All occurs equally*

* **Are there any missing values?**
* *Ans: No*

* **Which feature has the largest variation?**
* *Ans: PetalLengthCm*

# **Conclusion**


1. ***The dataset contains 150 flowers belonging to three species.***

2. ***Iris-virginica has the largest average petal length.***
1. ***There are no missing values.***
1. ***Petal measurements distinguish species better than sepal measurements.***
2. ***Pandas makes filtering, grouping, and statistical analysis much easier than NumPy.***




